pkg/manager/service.go

```diff
@@ -0,0 +1,528 @@
+func NewManager(ctx context.Context, syncAddress, diskUUID, diskPath, portRange string) (*Manager, error) {
+	workDir := filepath.Join(diskPath, types.BackingImageManagerDirectoryName)
+	if err := os.MkdirAll(workDir, 0666); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	start, end, err := ParsePortRange(portRange)
+	if err != nil {
+		return nil, err
+	}
+	m := &Manager{
+		ctx: ctx,
+
+		syncAddress: syncAddress,
+		diskUUID:    diskUUID,
+		diskPath:    diskPath,
+
+		portRangeMin:   start,
+		portRangeMax:   end,
+		availablePorts: util.NewBitmap(start, end),
+
+		lock:          &sync.RWMutex{},
+		biFileInfoMap: map[string]*api.FileInfo{},
+
+		broadcaster: &broadcaster.Broadcaster{},
+		broadcastCh: make(chan interface{}),
+
+		syncClient: &client.SyncClient{
+			Remote: syncAddress,
+		},
+
+		log: logrus.StandardLogger().WithFields(
+			logrus.Fields{
+				"component": "backing-image-manager",
+				"diskPath":  diskPath,
+				"diskUUID":  diskUUID,
+			},
+		),
+	}
+
+	// help to kickstart the broadcaster
+	if _, err := m.broadcaster.Subscribe(ctx, m.broadcastConnector); err != nil {
+		return nil, err
+	}
+	go m.monitoring()
+	go m.startBroadcasting(ctx)
+
+	return m, nil
+}
+
+func (m *Manager) startBroadcasting(ctx context.Context) {
+	ticker := time.NewTicker(types.MonitorInterval)
+	defer ticker.Stop()
+
+	done := false
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Info("Backing Image Manager: stopped broadcasting due to context done")
+			done = true
+			break
+		case <-ticker.C:
+			if m.checkBroadcasting() {
+				m.broadcastCh <- nil
+			}
+		}
+		if done {
+			break
+		}
+	}
+}
+
+func (m *Manager) checkBroadcasting() bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.broadcastRequired {
+		m.broadcastRequired = false
+		return true
+	}
+	return false
+}
+
+func (m *Manager) monitoring() {
+	ticker := time.NewTicker(types.MonitorInterval)
+	defer ticker.Stop()
+
+	done := false
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.log.Info("Backing Image Manager: stopped monitoring due to the context done")
+			done = true
+			break
+		case <-ticker.C:
+			m.listAndUpdate()
+		}
+		if done {
+			break
+		}
+	}
+}
+
+func (m *Manager) Delete(ctx context.Context, req *rpc.DeleteRequest) (resp *empty.Empty, err error) {
+	log := m.log.WithFields(logrus.Fields{"biName": req.Name, "biUUID": req.Uuid})
+	log.Info("Backing Image Manager: prepare to delete backing image")
+	defer func() {
+		if err != nil {
+			log.WithError(err).Error("Backing Image Manager: failed to delete backing image")
+		}
+	}()
+
+	if err := m.syncClient.Delete(types.GetBackingImageFilePath(m.diskPath, req.Name, req.Uuid)); err != nil {
+		return nil, err
+	}
+
+	if rmDirErr := os.RemoveAll(types.GetBackingImageDirectory(m.diskPath, req.Name, req.Uuid)); rmDirErr != nil {
+		log.Errorf("Backing Image Manager: failed to remove the backing image work directory after deleting the backing image file, will ignore it and continue: %v", rmDirErr)
+	}
```

Comment:
Would it be possible to delete the on-disk files again if the deletion failed?

Reply:
YES. Do you mean continuing executing this regardless of the above `m.syncClient.Delete` failure?

Reply:
Yes.